### PR TITLE
Implementation of (non-nightly) SIMD on FFM

### PR DESCRIPTION
**Pull request by ggaspersic**
* SIMD operations added to FFM where we do not need the nightly build.
* The biggest trick here is the additional unrolling of the loops, which gives us in total a roughly 15 % reduction on CPU & latency. 
* If we use SIMD vectors larger than 4 fields we start to get accuracy issues (hence the additional loop unrolling)

### CODE REVIEW

File: src/block_ffm.rs [hunk 1]
```diff
@@ -542,16 +554,53 @@ impl<L: OptimizerTrait + 'static> BlockTrait for BlockFFM<L> {
...
+                                contra_fields.as_mut_ptr().add(offset),
+                                field_embedding_len_as_usize,
+                            );
                         } else if !contra_fields_copied {
                             contra_fields_copied = true;
-                            for z in 0..field_embedding_len_as_usize {
+
+                            const LANES: usize = STEP * 4;
+                            let field_embedding_len_end = field_embedding_len_as_usize
+                                - (field_embedding_len_as_usize % LANES);
```
Comment (SkBlaz) [inline src/block_ffm.rs:567]: if lanes is 2**sth (it seems to be) you could further reduce the amount of instructions by & (LANES - 1)-ing it

File: src/block_ffm.rs [hunk 2]
```diff
@@ -542,16 +554,53 @@ impl<L: OptimizerTrait + 'static> BlockTrait for BlockFFM<L> {
...
+                            let mut cached_contra_fields_ptr =
+                                cached_contra_fields.as_ptr().add(offset);
+
+                            for _ in (0..field_embedding_len_end).step_by(LANES) {
+                                add_cached_contra_field(
+                                    contra_fields_ptr,
+                                    cached_contra_fields_ptr,
+                                );
+                                contra_fields_ptr = contra_fields_ptr.add(STEP);
+                                cached_contra_fields_ptr = cached_contra_fields_ptr.add(STEP);
```
Comment (SkBlaz) [inline src/block_ffm.rs:579]: What's the expected memory overhead of this cache btw?

File: src/block_ffm.rs [hunk 3]
```diff
@@ -879,49 +1074,91 @@ impl<L: OptimizerTrait + 'static> BlockFFM<L> {
...
+                    let contra_field_1 = _mm_loadu_ps(contra_fields_ptr_2);
+                    let acc_0 = _mm_mul_ps(contra_field_0, contra_field_1);
+
+                    let contra_field_2 = _mm_loadu_ps(contra_fields_ptr_1.add(STEP));
+                    let contra_field_3 = _mm_loadu_ps(contra_fields_ptr_2.add(STEP));
+                    let acc_1 = _mm_mul_ps(contra_field_2, contra_field_3);
+
+                    contra_field = hadd_ps(_mm_add_ps(acc_0, acc_1));
+                } else {
+                    for _ in (0..ffmk_end_as_usize).step_by(LANES) {
```
Comment (SkBlaz) [inline src/block_ffm.rs:1136]: If I understand this right, this is not 1:1 op map to previous version of loop structure right? These intermediary additions + the final normalization most likely result in different rig?

**Review by SkBlaz — COMMENTED** [review]
:zap:

File: src/block_ffm.rs [hunk 4]
```diff
@@ -542,16 +554,53 @@ impl<L: OptimizerTrait + 'static> BlockTrait for BlockFFM<L> {
...
+                            ptr::copy_nonoverlapping(
+                                cached_contra_fields.as_ptr().add(offset),
+                                contra_fields.as_mut_ptr().add(offset),
+                                field_embedding_len_as_usize,
+                            );
                         } else if !contra_fields_copied {
                             contra_fields_copied = true;
-                            for z in 0..field_embedding_len_as_usize {
+
+                            const LANES: usize = STEP * 4;
```
[the remark this answers is not in the document]
Comment (ggaspersic) [reply src/block_ffm.rs:565]: The accuracy  indeed drops already with the `f32x8`. With the 4 * 4 it actually allows for the interleaving of the command execution.

File: src/block_ffm.rs (hunk 1 above)
Comment (ggaspersic) [reply src/block_ffm.rs:567]: I do a step over everything as I need to move the pointer forward

File: src/block_ffm.rs (hunk 2 above)
Comment (ggaspersic) [reply src/block_ffm.rs:579]: There is no real overhead as its done per batch. & normally its a set of values for the number of cached features. So very small

File: src/block_ffm.rs (hunk 3 above)
Comment (ggaspersic) [reply src/block_ffm.rs:1136]: Actually it is the same. I only divided it more per use case where we have 8 & more (for future).
Created a new model with this binary based on another one & the rig was the same.

File: src/block_ffm.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (SkBlaz) [reply src/block_ffm.rs:565]: Ok thanks, makes  sense

File: src/block_ffm.rs (hunk 1 above)
Comment (SkBlaz) [reply src/block_ffm.rs:567]: Not sure I follow -> the suggestion was to get rid of modulo operator which on paper costs more than binary and-ing it appropriately

File: src/block_ffm.rs (hunk 2 above)
Comment (SkBlaz) [reply src/block_ffm.rs:579]: Right, this seems negligible w.r.t. weight space then

File: src/block_ffm.rs (hunk 3 above)
Comment (SkBlaz) [reply src/block_ffm.rs:1136]: Interesting, thanks for checking too!

**Review by SkBlaz — APPROVED** [review]

**Pull request merged**